### PR TITLE
Add Quant Row&Column ParallelLinear

### PR DESCRIPTION
**Pull request by RachelXu7**
### PR types

New features
### PR changes

APIs
### Describe

add QuantRowParallelLinear and QuantColumnParallelLinear in Quantization;
support model parallel QAT

### CODE REVIEW

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py [hunk 1]
```diff
@@ -48,7 +49,7 @@ class ImperativeQuantAware(object):
     """
 
     def __init__(self,
-                 quantizable_layer_type=['Conv2D', 'Linear', 'Conv2DTranspose'],
+                 quantizable_layer_type=['Conv2D', 'Linear', 'Conv2DTranspose', 'ColumnParallelLinear', 'RowParallelLinear'],
```
Comment (yghstill) [inline python/paddle/fluid/contrib/slim/quantization/imperative/qat.py:52]: quantizable_layer_type的默认值可不变，fleet定义QAT时传入'ColumnParallelLinear', 'RowParallelLinear'这些。

**Review by yghstill — COMMENTED** [review]
需要增加单测。

File: python/paddle/nn/quant/quant_layers.py [hunk 2]
```diff
@@ -658,14 +682,182 @@ def forward(self, input):
...
+        self.bias = getattr(layer, 'bias')
+        self.name = getattr(layer, '_name')
+        # For FakeQuant
+        self._linear_quant_axis = 1
+        
+        self.is_mp = getattr(layer, 'is_mp')
+        self.model_parallel_group = getattr(layer, 'model_parallel_group')
+        self.gather_output = getattr(layer, 'gather_output')
+        if weight_quant_layer is not None:
+            self._fake_quant_weight = weight_quant_layer()
```
Comment (wanghaoshuang) [inline python/paddle/nn/quant/quant_layers.py]: 当并行训练时，这里不允许自定义layer。需要给个warning或直接抛异常。

**Review by yghstill — DISMISSED** [review]
LGTM

**Review by XieYunshen — DISMISSED** [review]
LGTM 单测执行时间设置

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py [hunk 3]
```diff
@@ -29,6 +29,7 @@
 from paddle.fluid.initializer import Constant
 from paddle.fluid.dygraph.io import INFER_MODEL_SUFFIX, INFER_PARAMS_SUFFIX
 from paddle.fluid.io import load_inference_model, save_inference_model
+from paddle.distributed import fleet
```
Comment (qingqing01) [inline python/paddle/fluid/contrib/slim/quantization/imperative/qat.py]: 这个import没用到吧

File: python/paddle/fluid/tests/unittests/hybrid_parallel_qat.py [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+    def forward(self, x):
+        x = self.embedding(x)
+        x = self.linear1(x)
+        x = self.linear2(x)
+        x = self.linear3(x)
+        x = parallel_matmul(x, get_attr(self.embedding, "weight"), False)
+        return x
+
+
+class SimpleDPNet(fluid.dygraph.Layer):
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/hybrid_parallel_qat.py]: nn.Layer

File: python/paddle/nn/quant/quant_layers.py [hunk 5]
```diff
@@ -27,16 +28,11 @@
...
-    'QuantizedLinear',
-    'MovingAverageAbsMaxScale',
-    'MAOutputScaleLayer',
-    'FakeQuantMAOutputScaleLayer',
-    'QuantStub',
+    'FakeQuantAbsMax', 'FakeQuantMovingAverageAbsMax',
+    'FakeQuantChannelWiseAbsMax', 'QuantizedConv2D', 'QuantizedConv2DTranspose',
+    'QuantizedLinear', 'MovingAverageAbsMaxScale', 'MAOutputScaleLayer',
+    'FakeQuantMAOutputScaleLayer', 'QuantStub', 'QuantizedRowParallelLinear',
+    'QuantizedColumnParallelLinear'
```
Comment (qingqing01) [inline python/paddle/nn/quant/quant_layers.py]: 最后加个逗号，pre-commit格式化下(一行一个，不用手动一行行改)~

File: python/paddle/nn/quant/quant_layers.py [hunk 6]
```diff
@@ -58,10 +54,12 @@ def __init__(self,
                  name=None,
                  quant_bits=8,
                  dtype='float32',
-                 quant_on_weight=False):
+                 quant_on_weight=False,
+                 reduce_type=None):
```
Comment (qingqing01) [inline python/paddle/nn/quant/quant_layers.py:65]: add comments for  all new arguments `reduce_type`

File: python/paddle/nn/quant/quant_layers.py [hunk 7]
```diff
@@ -659,13 +681,184 @@ def forward(self, input):
...
+                 layer,
+                 weight_bits=8,
+                 activation_bits=8,
+                 moving_rate=0.9,
+                 weight_quantize_type='abs_max',
+                 activation_quantize_type='abs_max',
+                 weight_pre_layer=None,
+                 act_pre_layer=None,
+                 weight_quant_layer=None,
+                 act_quant_layer=None):
```
Comment (qingqing01) [inline python/paddle/nn/quant/quant_layers.py:703]: add comments for all input args

File: python/paddle/fluid/tests/unittests/hybrid_parallel_qat.py [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+import paddle
+import numpy as np
+import random
+import paddle.distributed as dist
+import paddle.fluid as fluid
+import paddle.distributed.fleet as fleet
+from paddle.io import DataLoader, Dataset
+import unittest
+import paddle.nn as nn
+from paddle.fluid.contrib.slim.quantization import ImperativeQuantAware
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/hybrid_parallel_qat.py:28]: better to put this unit testing into quantization testing directory

**Review by qingqing01 — APPROVED** [review]

**Review by XieYunshen — APPROVED** [review]

**Review by dingjiaweiww — APPROVED** [review]

**Review by yghstill — APPROVED** [review]

**Pull request merged**